File: python/string/canConstruct.py
#!/usr/bin/env python3
#coding=utf-8

import collections

''' leetcode issue 383'''
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        c1 = collections.Counter(magazine)
        c2 = collections.Counter(ransomNote)
        return c2 - c1
        '''
        sup = [0] * 26
        for c in magazine:
            sup[ord(c) - ord('a')] += 1
        for c in ransomNote:
            if(sup[ord(c) - ord('a')]):
                sup[ord(c) - ord('a')] -= 1
            else:
                return False
        return True
        '''

File: c++/string/canConstruct.cc
#include <string>
#include <vector>

using namespace std;

/* leetcode issue 383 */
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        vector<int> sup(26, 0);

        for(char c : magazine) {
            sup[c-'a']++;
        }

        for(char c: ransomNote) {
            if(sup[c-'a'] > 0) {
                sup[c-'a']--;
            } else {
                return false;
            }
        }

        return true;
    }
};
